proposals/pattern-based-fixed.md

```diff
@@ -0,0 +1,135 @@
+-	Convey unambiguously the type of the unmanaged element   (i.e. �char� for �string�)
+-	Prescribe the behavior in "empty" case when there is nothing to refer to. 
+-	Should not push API authors toward design decisions that hurt the use of the type outside of `fixed`.
+
+
+I think the above could be satisfied by recognizing a specially named ref-returning member:
+ `ref [readonly] T DangerousGetPinnableReference()`.
+
+In order to be used by the `fixed` statement the following conditions must be met:
+
+1)	There is only one such member provided for a type.
+1)	Returns by `ref` or `ref readonly`. 
+(readonly is permitted so that authors of immutable/readonly types could implement the pattern)
+1)  T is an unmanaged type.
+(since `T*` becomes the pointer type)
+1)	Returns managed `nullptr` when there is no data to pin � probably the cheapest way to convey emptiness.
+(note that �� string returns a ref to �\0� since strings are zero-terminated)
+
+
+Alternatively for the `#3` we can allow the result in empty cases be undefined or implementation-specific. 
+That, however, may make the API more dangerous and prone to abuse and unintended compatibility burdens. 
+
+## *Translation* ##
+
+
+```C#
+    fixed(byte* ptr = thing)
+    { 
+		// <BODY>
+    }
+```
+
+becomes the following pseudocode (not all expressible in C#)
+
+```C#
+
+    byte* ptr;
+    // specially decorated "pinned" IL local slot, not visible to user code.
+    pinned ref byte _pinned;
+
+    try
+    {
+        // NOTE: null check is omitted for value types 
+        // NOTE: `thing` is evaluated only once (temporary is introduced if necessary) 
+        if (thing != null)
+        {
+            // obtain and "pin" the reference
+            _pinned = ref thing.DangerousGetPinnableReference();
+
+            // unsafe cast in IL
+            ptr = (byte*)_pinned;
+        }
+        else
+        {
+            ptr = default(byte*);
+        }
+
+        // <BODY> 
+    }
+    finally   // finally can be omitted when not observable
+    {
+        // "unpin" the object
+        _pinned = nullptr;
+    }
+
+```
+
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+- DangerousGetPinnableReference is intended to be used only in `fixed`, but nothing prevents its use in safe code, so implementor must keep that in mind.
+
+## Alternatives
+[alternatives]: #alternatives
+
+Users can introduce DangerousGetPinnableReference or similar member and use it as
+ 
+```C#
+    fixed(byte* ptr = thing.DangerousGetPinnableReference())
+    { 
+		// <BODY>
+    }
+```
+
+There is no solution for `System.String` if alternative solution is desired.
+
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+- [ ] Behavior in "empty" state. - `nullptr` or `undefined` ? 
+- [ ] Should the extension methods be considered ? 
+- [ ] If a pattern is detected on `System.String`, should it win over ? 
```

Comment:
Yes!

Reply:
Also think so. Otherwise - what is the point?